runtime/freelist.c

```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+        caml_fl_cur_wsz += Whsize_wosize (least_wosz) - Whsize_wosize (wosz);
+          /* remnant stays in tree */
+        CAMLassert (Color_val ((value) bf_large_least) == Caml_blue);
+        return result;
+      }
+    }
+
+    /* Allocate from the tree and update [bf_large_least]. */
+    result = bf_allocate_from_tree (wosz, 1);
+    FREELIST_DEBUG_bf_check ();
+    return result;
+  }else{
+    result = bf_allocate_from_tree (wosz, 0);
+    FREELIST_DEBUG_bf_check ();
+    return result;
+  }
+}
+
+static void bf_init_merge (void)
+{
+  mlsize_t i;
+
+#ifdef CAML_INSTR
+  for (i = 1; i < 20; i++){
+    CAML_INSTR_INT (instr_name[i], instr_size[i]);
+    instr_size[i] = 0;
+  }
+#endif /* CAML_INSTR */
+
+  caml_fl_merge = Val_NULL;
+
+  for (i = 1; i <= BF_NUM_SMALL; i++){
+    /* At the beginning of each small free list is a segment of remnants
+       that were pushed back to the list after splitting. These are white
+       and they are not in order. We need to remove them
+       from the list for coalescing to work. They
+       will be picked up by the sweeping code and inserted in the right
+       place in the list.
+    */
+    value p = bf_small_fl[i].free;
+    while (1){
+      if (p == Val_NULL){
+        unset_map (i);
+        break;
+      }
+      if (Color_val (p) == Caml_blue) break;
+      CAMLassert (Color_val (p) == Caml_white);
+      caml_fl_cur_wsz -= Whsize_val (p);
+      p = Next_small (p);
+    }
+    bf_small_fl[i].free = p;
+    /* Set the merge pointer to its initial value */
+    bf_small_fl[i].merge = &bf_small_fl[i].free;
+  }
+}
+
+static void bf_reset (void)
+{
+  mlsize_t i;
+
+  for (i = 1; i <= BF_NUM_SMALL; i++){
+    bf_small_fl[i].free = Val_NULL;
+    bf_small_fl[i].merge = &(bf_small_fl[i].free);
+  }
+  bf_small_map = 0;
+  bf_large_tree = NULL;
+  bf_large_least = NULL;
+  caml_fl_cur_wsz = 0;
+  bf_init_merge ();
+}
+
+static header_t *bf_merge_block (value bp, char *limit)
+{
+  value start;
+  value cur;
+  mlsize_t wosz;
+
+  CAMLassert (Color_val (bp) == Caml_white);
+  /* Find the starting point of the current run of free blocks. */
+  if (caml_fl_merge != Val_NULL && Next_in_mem (caml_fl_merge) == bp){
+    start = caml_fl_merge;
+    CAMLassert (Color_val (start) == Caml_blue);
+    bf_remove (start);
+  }else{
+    start = bp;
+  }
+  cur = bp;
+  while (1){
+    /* This slightly convoluted loop is just going over the run of
+       white or blue blocks, doing the right thing for each color, and
+       stopping on a gray or black block or when limit is passed.
+       It is convoluted because we start knowing that the first block
+       is white. */
+  white:
+    if (Tag_val (cur) == Custom_tag){
+      void (*final_fun)(value) = Custom_ops_val(cur)->finalize;
+      if (final_fun != NULL) final_fun(cur);
+    }
+    caml_fl_cur_wsz += Whsize_val (cur);
+  next:
+    cur = Next_in_mem (cur);
+    if ((char *) cur > limit) goto end_of_run;
```

Comment:
Isn't it `cur >= limit` ?

Reply:
cur points to the value rather than the header, so I think `cur > limit` is correct. Iff we have a size zero remnant at the end of the heap (I don't know whether that is actually possible) then we would see `cur = limit` in which case we should continue to merge it onto the current block.

Reply:
Ah, you are right! God, this is misleading!

Reply:
Perhaps we could write:
```suggestion
    if (cur >= Val_hp(limit)) goto end_of_run;
```
to ake thi clearer?

Reply:
I changed it to `Hp_val(cur) >= limit` to make it homogeneous with line 558 of major_gc.c.
(and less of an undefined behavior)